internal/server/apimigrator.go

```diff
@@ -26,24 +26,48 @@ type apiMigration struct {
 	redirect        string
 	requestRewrite  func(c *gin.Context)
 	responseRewrite func(c *gin.Context)
+	redirectHandler func(c *gin.Context)
 }
 
-func addRedirect(a *API, method, path, newPath, version string) {
+func addRedirect(a *API, method, oldPath, newPath, version string) {
+	migration := apiMigration{
+		method:   method,
+		path:     oldPath,
+		version:  version,
+		redirect: newPath,
+	}
+
+	// a closure, so that it uses the updated path when it runs
+	migration.redirectHandler = func(c *gin.Context) {
+		fmt.Printf("Applying redirect path=%v from=%v to=%v\n",
+			c.Request.URL.Path, migration.path, migration.redirect)
+		c.Request.URL.Path = migration.redirect
+		c.Next()
+	}
+
+	updateRedirectMigrations(a, migration)
+	a.migrations = append(a.migrations, migration)
+}
+
+// updateRedirectMigrations updates the a.migrations list so that any existing
+// migrations are applied to the new canonical path for the route.
+func updateRedirectMigrations(a *API, newMigration apiMigration) {
 	// update any existing migrations with the legacy path
 	for i, mig := range a.migrations {
-		if len(mig.redirect) == 0 && mig.path == path {
-			a.migrations[i].path = newPath
+
+		// TODO: removing this causes test to fail, but this does not seem right.
+		// It causes a bunch of incorrect rewrites to be applied to the new migration.
+		if mig.redirect == "" && mig.path == newMigration.path {
+			fmt.Println("Updating ", a.migrations[i].path, " to ", newMigration.redirect)
+			a.migrations[i].path = newMigration.redirect
 		}
-		if len(mig.redirect) > 0 && mig.redirect == path {
-			a.migrations[i].redirect = newPath
+		// Update any existing redirect migrations, so that the target of the redirect
+		// is the new canonical path.
+		// TODO: but what if there are intermediate paths that are necessary?
+		if mig.redirect != "" && mig.redirect == newMigration.path {
+			a.migrations[i].redirect = newMigration.redirect
```

Comment:
I think this update logic (previously it was in `addRedirect`) is not quite right. It seems to work when there is only one migration for the endpoint, but adding another one in this PR causes some strange behaviour.

The tests pass, but if you look at the debugging output I added "Applying ...", it applies a bunch of request rewrites incorrectly. 

It doesn't seem correct to update all of these migrations because it changes the meaning of them.

Reply:
Reverted this part. The problem is more involved than simply this update.